src/System.Private.ServiceModel/src/System/IdentityModel/Tokens/X509AsymmetricSecurityKey.cs

```diff
@@ -0,0 +1,450 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using System.Security.Cryptography.Xml;
+using System.ServiceModel;
+
+namespace System.IdentityModel.Tokens
+{
+    public class X509AsymmetricSecurityKey : AsymmetricSecurityKey
+    {
+        private X509Certificate2 _certificate;
+        private AsymmetricAlgorithm _privateKey;
+        private bool _privateKeyAvailabilityDetermined;
+        private AsymmetricAlgorithm _publicKey;
+        private bool _publicKeyAvailabilityDetermined;
+
+        private object _thisLock = new Object();
+
+        public X509AsymmetricSecurityKey(X509Certificate2 certificate)
+        {
+            _certificate = certificate ?? throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(certificate));
+        }
+
+        public override int KeySize
+        {
+            get { return this.PublicKey.KeySize; }
+        }
+
+        private AsymmetricAlgorithm PrivateKey
+        {
+            get
+            {
+                if (!_privateKeyAvailabilityDetermined)
+                {
+                    lock (ThisLock)
+                    {
+                        _privateKey = _certificate.GetRSAPrivateKey();
+                        if (_privateKey != null)
+                        {
+                            RSACryptoServiceProvider rsaCsp = _privateKey as RSACryptoServiceProvider;
+                            // ProviderType == 1 is PROV_RSA_FULL provider type that only supports SHA1. Change it to PROV_RSA_AES=24 that supports SHA2 also.
+                            if (rsaCsp != null && rsaCsp.CspKeyContainerInfo.ProviderType == 1)
+                            {
+                                CspParameters csp = new CspParameters();
+                                csp.ProviderType = 24;
+                                csp.KeyContainerName = rsaCsp.CspKeyContainerInfo.KeyContainerName;
+                                csp.KeyNumber = (int)rsaCsp.CspKeyContainerInfo.KeyNumber;
+                                if (rsaCsp.CspKeyContainerInfo.MachineKeyStore)
+                                    csp.Flags = CspProviderFlags.UseMachineKeyStore;
+
+                                csp.Flags |= CspProviderFlags.UseExistingKey;
+                                _privateKey = new RSACryptoServiceProvider(csp);
+                            }
+                        }
+                        else
+                        {
+                            throw ExceptionHelper.PlatformNotSupported();
+                            // Need DSACertificateExtensions which is in netstandard2.1 and netcore2.0. As we target netstandard2.0, we don't
+                            // have access to DSACertificateExtensions 
```

Comment:
No ECDSA support?

Reply:
There isn't any on .NET Framework so I didn't add it. I'll add it here.

---
In reply to: [274016549](https://github.com/dotnet/wcf/pull/3538#discussion_r274016549) [](ancestors = 274016549)